go/login/server/server.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package backend implements GLOME-login server framework.
+package server
+
+import (
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+
+	"github.com/google/glome/go/glome"
+	"github.com/gorilla/mux"
+)
+
+const (
+	// MaxResponseSize is the maximum size in characaters of the response token
+	MaxResponseSize = 44
+)
+
+var (
+	// ErrInvalidResponseLen  denotes that a ephemeral is invalid.
+	ErrInvalidResponseLen = fmt.Errorf("invalid response length provided")
+
+	// This error are hypothesis on what will be provided by the glome login lib
+
+	// ErrFailedHandshake denotes that GLOME handshake has failed.
+	ErrFailedHandshake = fmt.Errorf("400 bad request: failed handshake")
+	// ErrVersionNotSupported denotes that a version of GLOME handshake is not supported.
+	ErrVersionNotSupported = fmt.Errorf("400 bad request: version not supported")
+	// ErrInvalidPrefixType denotes that a prefix type is invalid.
+	ErrInvalidPrefixType = fmt.Errorf("400 bad request: invalid prefix type")
+	// ErrInvalidPrefix7 denotes that a prefix7 is invalid.
+	ErrInvalidPrefix7 = fmt.Errorf("404 Not Found: prefix7 not found")
+	// ErrInvalidEph denotes that a ephemeral is invalid.
+	ErrInvalidEph = fmt.Errorf("400 bad request: invalid ephemeral Key")
+	// ErrIncorrectTag denotes that received tag is incorrect
+	ErrIncorrectTag = fmt.Errorf("invalid Tag")
+)
+
+// Authorizer responds to an authorization request. The method
+// Allow returns whether an user is allowed to run an action in a host.
+//
+// Some considerations need to be held while implementing this interface:
+// - Allow should accept an empty string as command, denoting that
+// an empty action is passed in the request.
+// - If no user can be obtained from request metadata, an empty string is to be
+// passed as default value.
+// - The host can contain information about both the host type and the host id parsed
+// as <host-type>:<host-id>. the host-type information is optional. if host-type is not
+// provided the string is parsed as <host-id> (that is, we also drop the colon).
+type Authorizer interface {
+	GrantLogin(user string, hostID string, hostIDType string, action string) bool
```

Comment:
It might be useful to convey a reason for the authorization decision. In some systems that reason may be obvious, but a user that is being rejected may want to know whether this is working-as-intended or whether there was a transient error in a dependent service that prevented an authorization decision.

Reply:
SGTM!